--- firefox-96.0/python/mach/mach/site.py.orig	2022-01-06 19:32:46.000000000 +0100
+++ firefox-96.0/python/mach/mach/site.py	2022-01-12 13:21:03.273897180 +0100
@@ -848,10 +848,7 @@
 
 
 def _virtualenv_py_path(topsrcdir):
-    return os.path.join(
-        topsrcdir, "third_party", "python", "virtualenv", "virtualenv.py"
-    )
-
+    return '/usr/bin/virtualenv-3'
 
 def _resolve_installed_packages(python_executable):
     pip_json = subprocess.check_output(
